modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/util/StreamInterceptorsLoader.java

```diff
@@ -0,0 +1,158 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.synapse.transport.passthru.util;
+
+import org.apache.axiom.om.OMElement;
+import org.apache.axiom.om.OMNode;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.commons.util.MiscellaneousUtil;
+import org.apache.synapse.commons.util.PropertyHelper;
+import org.apache.synapse.transport.passthru.StreamInterceptor;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import javax.xml.namespace.QName;
+
+public class StreamInterceptorsLoader {
+
+    StreamInterceptorsLoader() {
+    }
+
+    private static Log log = LogFactory.getLog(StreamInterceptorsLoader.class);
+
+    private static final QName ROOT_Q = new QName("interceptors");
+    private static final QName INTERCEPTOR_Q = new QName("interceptor");
+    private static final QName CLASS_Q = new QName("class");
+    private static final QName NAME_ATT = new QName("name");
+    private static final QName PARAM_Q = new QName("parameter");
+    private static final QName VALUE_ATT = new QName("value");
+
+    private static List<StreamInterceptor> interceptors = new ArrayList<>();
+
+    private static boolean isLoadedAlready = false;
+
+    /**
+     * Load and get all synapse interceptors
+     *
+     * @return List of loaded synapse interceptors
+     */
+    public static List<StreamInterceptor> getInterceptors() {
+
+        if (!isLoadedAlready) {
+            loadInterceptors();
+            isLoadedAlready = true;
+        }
+        return Collections.unmodifiableList(interceptors);
+    }
+
+    private static void loadInterceptors() {
+
+        OMElement interceptorsConfig = MiscellaneousUtil.loadXMLConfig(RelayConstants.STREAM_INTERCEPTOR_FILE);
+        if (interceptorsConfig != null) {
+
+            if (!ROOT_Q.equals(interceptorsConfig.getQName())) {
+                handleException("Invalid interceptor configuration file");
+            }
+
+            Iterator iterator = interceptorsConfig.getChildrenWithName(INTERCEPTOR_Q);
+            while (iterator.hasNext()) {
+                OMElement interceptorElem = (OMElement) iterator.next();
+
+                String name = null;
+                if (interceptorElem.getAttribute(NAME_ATT) != null) {
+                    name = interceptorElem.getAttributeValue(NAME_ATT);
+                } else {
+                    handleException("Name not defined in one or more interceptor");
+                }
+
+                if (interceptorElem.getAttribute(CLASS_Q) != null) {
+                    String className = interceptorElem.getAttributeValue(CLASS_Q);
+                    if (!"".equals(className)) {
+                        StreamInterceptor interceptor = createInterceptor(className);
+                        interceptors.add(interceptor);
+                        interceptor.setName(name);
```

Comment:
To me, it seems as if the name cannot change in the runtime once the interceptor is created. Why don't we remove the setter and add the parameter to the constructor itself?

Reply:
added a lock around this